#121 #129 API for create & list WIT

pranavgore09 opened the pull request:
REST api for Create WIT.
Accepts simple json input. Creates only Simple Type as of now.
Example is added into examples/create_workitemtype.go

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/almighty/almighty-core/149)
<!-- Reviewable:end -->

pranavgore09 commented:
`make tes-all` fails as of now. I need to check yet.

kwk commented:
@pranavgore09 Try setting the `ALMIGHTY_DB_HOST` environment variable to point to your DB:

```sh
ALMIGHTY_DB_HOST=172.0.0.1 make test-all
```

I need to update the documentation.

pranavgore09 commented:
yeah, thanks. this too worked. But had to make some code changes to add fixes for failing tests.

tsmaeder commented:
Reviewed 1 of 4 files at r1, 2 of 9 files at r2, 2 of 2 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[workitemtype.go, line 51 \[r5\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPrhwB-8JQqhro6_f0c:-KPrhwB-8JQqhro6_f0d:bwakcvb) ([raw file](https://github.com/almighty/almighty-core/blob/7f18c31ef9ff22d18b248de834db01f9ff354d04/workitemtype.go#L51)):*
> ```Go
> 
> 		// // need to convert field value to app.FieldDefinition
> 		for key, value := range ctx.Payload.Fields {
> ```

You can get rid of this conversion loop if you update payload definition to use HashOf(string, fieldDefinition). The fieldDefintion type is available if you rebase your patch on the latest state of the branch.

---

*[workitemtype_test.go, line 13 \[r5\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPriaJ0iLu8xFl17fvz:-KPriaJ0iLu8xFl17fw-:bdodo8x) ([raw file](https://github.com/almighty/almighty-core/blob/7f18c31ef9ff22d18b248de834db01f9ff354d04/workitemtype_test.go#L13)):*
> ```Go
> 
> func TestGetWorkItemType(t *testing.T) {
> 	t.Skip("work in progress hence not testing now.")
> ```

Perhaps it makes sense to implement "delete" in the WorkitemTypeRepository and update this test to create/get/delete a wit each time.
Does golang have a "todo" tag?

---

*[design/resources.go, line 130 \[r5\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPrj1-1mOdxppL82vIL:-KPrj1-1mOdxppL82vIM:b9mxkcx) ([raw file](https://github.com/almighty/almighty-core/blob/7f18c31ef9ff22d18b248de834db01f9ff354d04/design/resources.go#L130)):*
> ```Go
> 		})
> 		Response(InternalServerError)
> 		Response(NotFound)
> ```

I don't think 404 makes sense here. 

---

*[design/user_types.go, line 26 \[r5\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPrgzlzMkEEwldXnyVe:-KPrgzm-7dzavI3mUi2a:bgzmysg) ([raw file](https://github.com/almighty/almighty-core/blob/7f18c31ef9ff22d18b248de834db01f9ff354d04/design/user_types.go#L26)):*
> ```Go
> })
> 
> var CreateWorkItemTypePayload = Type("CreateWorkItemTypePayload", func() {
> ```

Since you created this, I have pushed new Goa type definitions for "fieldDefinition". I would make sense to reuse those in order to keep the API consistent. 
    Attribute("fields", HashOf(String, fieldType), "Type fields those must be followed by respective Work Items.")

---

*[design/user_types.go, line 29 \[r5\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPrjJ-QgZLM4inYJVOW:-KPrjJ-QgZLM4inYJVOX:b-21lxt8) ([raw file](https://github.com/almighty/almighty-core/blob/7f18c31ef9ff22d18b248de834db01f9ff354d04/design/user_types.go#L29)):*
> ```Go
> 	Attribute("name", String, "Readable name of the type like Task, Issue, Bug, Epic etc.")
> 	Attribute("fields", HashOf(String, Any), "Type fields those must be followed by respective Work Items.")
> 	Attribute("extendedTypeID", String, "If newly created type extends any existing type")
> ```

Should be "extendedTypeName", really, no?

---

*[examples/create_workitemtype.json, line 13 \[r5\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPrjNalD6JTFET501Gm:-KPrjNalD6JTFET501Gn:b-tmxhx1) ([raw file](https://github.com/almighty/almighty-core/blob/7f18c31ef9ff22d18b248de834db01f9ff354d04/examples/create_workitemtype.json#L13)):*
> ```JSON
> ' -H localhost:8080 --pp
> 
> ./bin/alm-cli show workitemtype --name "Issue" -H localhost:8080 --pp
> ```

Examples are a nice touch. 

---

*[workitem_test.go, line 2 \[r4\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPrjisyDOivrj4YCGvG:-KPrjisyDOivrj4YCGvH:bz221cl) ([raw file](https://github.com/almighty/almighty-core/blob/39b352ef23f36d7e6ada347a77b80380e9669267/workitem_test.go#L2)):*
> ```Go
> // +build integration
> ```

We should keep this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

pranavgore09 commented:
Review status: 0 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[workitem_test.go, line 2 \[r4\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPrjisyDOivrj4YCGvG:-KQ-tdMJKwAFnVL1U9NH:bq70vcj) ([raw file](https://github.com/almighty/almighty-core/blob/39b352ef23f36d7e6ada347a77b80380e9669267/workitem_test.go#L2)):*
<details><summary><i>Previously, tsmaeder (Thomas Mäder) wrote…</i></summary>
> We should keep this.
</details>
yup accidently removed. added back in recent commit.

---

*[workitemtype.go, line 51 \[r5\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPrhwB-8JQqhro6_f0c:-KQ-syD4fbi7J4GT2w9W:b3wcfwr) ([raw file](https://github.com/almighty/almighty-core/blob/7f18c31ef9ff22d18b248de834db01f9ff354d04/workitemtype.go#L51)):*
<details><summary><i>Previously, tsmaeder (Thomas Mäder) wrote…</i></summary>
> You can get rid of this conversion loop if you update payload definition to use HashOf(string, fieldDefinition). The fieldDefintion type is available if you rebase your patch on the latest state of the branch.
</details>
yeah, I have updated to use HashOf(string, fieldDefinition). But essentially I need to loop through to dereference every fieldDefinition object.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

pranavgore09 commented:
Review status: 0 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[design/resources.go, line 130 \[r5\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPrj1-1mOdxppL82vIL:-KQ-vakPznjhsVDbeE9D:b-bnwjrb) ([raw file](https://github.com/almighty/almighty-core/blob/7f18c31ef9ff22d18b248de834db01f9ff354d04/design/resources.go#L130)):*
<details><summary><i>Previously, pranavgore09 (Pranav Gore) wrote…</i></summary>
> yes 404 for Create call, my bad. Thanks for pointing out.
</details>
done, removed as needed for create action. And we need it for delete action.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

aslakknutsen commented on models/workitemtype_repository.go:
```diff
@@ -101,6 +101,26 @@ func (r *GormWorkItemTypeRepository) Create(ctx context.Context, extendedTypeNam
 	return &result, nil
 }
 
+// Delete deletes the work item with the given id
```
"work item type"

aslakknutsen commented on workitemtype.go:
```diff
@@ -41,3 +41,42 @@ func (c *WorkitemtypeController) Show(ctx *app.ShowWorkitemtypeContext) error {
+}
+
+// Delete runs the delete action.
+func (c *WorkitemtypeController) Delete(ctx *app.DeleteWorkitemtypeContext) error {
```
What happens to the WI's that depend on this WIT?

pranavgore09 commented:
Review status: 0 of 10 files reviewed at latest revision, 9 unresolved discussions.

---

*[models/workitemtype_repository.go, line 104 \[r6\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqpv2vKrm9GfBoC8iq-r6-104:-KQ-zJEyu1R_k9vZy7f6:bksy3ka) ([raw file](https://github.com/almighty/almighty-core/blob/6594b636cf10dce5a632756442d824a8350645dc/models/workitemtype_repository.go#L104)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> "work item type"
</details>
yes, I will change to
"deletes the work item type with the given name"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

kwk commented:
Reviewed 10 of 10 files at r6.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[workitemtype.go, line 69 \[r6\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmo-r6-69:-KQ00YzPnAuAcoKwvWTK:b-6gz5jm) ([raw file](https://github.com/almighty/almighty-core/blob/6594b636cf10dce5a632756442d824a8350645dc/workitemtype.go#L69)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> What happens to the WI's that depend on this WIT?
</details>
Very good point indeed. I think there needs to be some user interaction involved as well. For example, ask the user if all WI's should be deleted as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

pranavgore09 commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[workitemtype.go, line 69 \[r6\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmo-r6-69:-KQ0-cxZ0lWPOTSrLsMi:bqjimpy) ([raw file](https://github.com/almighty/almighty-core/blob/6594b636cf10dce5a632756442d824a8350645dc/workitemtype.go#L69)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> What happens to the WI's that depend on this WIT?
</details>
Frankly, I missed this point. Sorry.
My understanding was, UI will ask user to perform action. If user says delete all related all items then we can clear them. Else we can change their types to some other type(suggested by user). Those items who fail in transition because of constraints will be shown to user for further action. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

pranavgore09 commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[workitemtype.go, line 69 \[r6\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmo-r6-69:-KQ017H4biF4T0r0vd7b:bnc9uit) ([raw file](https://github.com/almighty/almighty-core/blob/6594b636cf10dce5a632756442d824a8350645dc/workitemtype.go#L69)):*
<details><summary><i>Previously, pranavgore09 (Pranav Gore) wrote…</i></summary>
> Frankly, I missed this point. Sorry.
> My understanding was, UI will ask user to perform action. If user says delete all related all items then we can clear them. Else we can change their types to some other type(suggested by user). Those items who fail in transition because of constraints will be shown to user for further action. 
</details>
yes Konrad.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

kwk commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[workitemtype.go, line 69 \[r6\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmo-r6-69:-KQ01EkTzf4B8r0-9uLW:b-xcq46j) ([raw file](https://github.com/almighty/almighty-core/blob/6594b636cf10dce5a632756442d824a8350645dc/workitemtype.go#L69)):*
<details><summary><i>Previously, pranavgore09 (Pranav Gore) wrote…</i></summary>
> yes Konrad.
</details>
For now I'd say that it is okay to delete WIT's with no WI's that depend on it. Otherwise raise a HTTP error:
e.g. 409 Conflict, or 424 Failed Dependency, or 428 Precondition Required. Whatever action is taken an appropriate response must be added to the delete resource's action.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

pranavgore09 commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[workitemtype.go, line 69 \[r6\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmo-r6-69:-KQ0269u2E7hOj5XiUa7:b-7wu2fc) ([raw file](https://github.com/almighty/almighty-core/blob/6594b636cf10dce5a632756442d824a8350645dc/workitemtype.go#L69)):*
<details><summary><i>Previously, kwk (Konrad Kleine) wrote…</i></summary>
> For now I'd say that it is okay to delete WIT's with no WI's that depend on it. Otherwise raise a HTTP error:
> e.g. 409 Conflict, or 424 Failed Dependency, or 428 Precondition Required. Whatever action is taken an appropriate response must be added to the delete resource's action.
</details>
Should I setup a way to perform cascade delete for respective WIs ? Like a flag , if True, it will trigger cascade delete.
And also "Type" is not foreign key in WI. If we make it foreign key then GORM can take care of this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

pranavgore09 commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[workitemtype.go, line 69 \[r6\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmo-r6-69:-KQ02jtcPlxUL9FB86hi:b-ni7s5t) ([raw file](https://github.com/almighty/almighty-core/blob/6594b636cf10dce5a632756442d824a8350645dc/workitemtype.go#L69)):*
<details><summary><i>Previously, pranavgore09 (Pranav Gore) wrote…</i></summary>
> Should I setup a way to perform cascade delete for respective WIs ? Like a flag , if True, it will trigger cascade delete.
> And also "Type" is not foreign key in WI. If we make it foreign key then GORM can take care of this.
</details>
Yeah Konrad, that also makes sense to me. !
if no related WI then delete
else raise 4**()..

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

aslakknutsen commented:
*[workitemtype.go, line 69 \[r6\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmo-r6-69:-KQ046rspBHllNXx6SHF:b4f4rb0) ([raw file](https://github.com/almighty/almighty-core/blob/6594b636cf10dce5a632756442d824a8350645dc/workitemtype.go#L69)):*
<details><summary><i>Previously, pranavgore09 (Pranav Gore) wrote…</i></summary>
> Yeah Konrad, that also makes sense to me. !
> if no related WI then delete
> else raise 4**()..
</details>
We should never actually hard delete anything. 

As we don't really have any clear idea what a DELETE WIT actually means, I'd say remove it from the API.

We're not making a raw CRUD REST API on top of a DB. Our API has to reflect how the 'business workflow' is intended to work... Without a clear understanding of what the process is, don't implement/expose the feature.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

pranavgore09 commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[workitemtype.go, line 69 \[r6\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmo-r6-69:-KQ06HbTzYqLKiRK9g2Y:bp7mj54) ([raw file](https://github.com/almighty/almighty-core/blob/6594b636cf10dce5a632756442d824a8350645dc/workitemtype.go#L69)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> We should never actually hard delete anything. 
> 
> As we don't really have any clear idea what a DELETE WIT actually means, I'd say remove it from the API.
> 
> We're not making a raw CRUD REST API on top of a DB. Our API has to reflect how the 'business workflow' is intended to work... Without a clear understanding of what the process is, don't implement/expose the feature.
> 
</details>
Right that too.
I can remove the code that exposes the API endpoint only (design/resources.go). And rest of the code can be there. It will be used to remove WIT in TearDown of tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

pranavgore09 commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[workitemtype.go, line 69 \[r6\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmo-r6-69:-KQ08XIa2EuV7xj211aX:blx4o1e) ([raw file](https://github.com/almighty/almighty-core/blob/6594b636cf10dce5a632756442d824a8350645dc/workitemtype.go#L69)):*
<details><summary><i>Previously, pranavgore09 (Pranav Gore) wrote…</i></summary>
> Right that too.
> I can remove the code that exposes the API endpoint only (design/resources.go). And rest of the code can be there. It will be used to remove WIT in TearDown of tests.
> 
</details>
Ok, one more related question. "API for Update WIT". What to do with existing WIs of that type. Do we have workflow for that ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

kwk commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[workitemtype.go, line 69 \[r6\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmo-r6-69:-KQ09QaUWBsp9onjlRS_:btvkv5i) ([raw file](https://github.com/almighty/almighty-core/blob/6594b636cf10dce5a632756442d824a8350645dc/workitemtype.go#L69)):*
<details><summary><i>Previously, pranavgore09 (Pranav Gore) wrote…</i></summary>
> Ok, one more related question. "API for Update WIT". What to do with existing WIs of that type. Do we have workflow for that ?
</details>
@aslakknutsen this approach reminds me of what Mitch was saying in the sos yesterday. Removing the Delete capability is more experience driven rather than a generic solution. I therefore fully agree to removing it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

aslakknutsen commented:
*[workitemtype.go, line 69 \[r6\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmo-r6-69:-KQ0HIZzTvc9iD8Xvjks:bcsbjj2) ([raw file](https://github.com/almighty/almighty-core/blob/6594b636cf10dce5a632756442d824a8350645dc/workitemtype.go#L69)):*
<details><summary><i>Previously, kwk (Konrad Kleine) wrote…</i></summary>
> @aslakknutsen this approach reminds me of what Mitch was saying in the sos yesterday. Removing the Delete capability is more experience driven rather than a generic solution. I therefore fully agree to removing it.
</details>
@pranavgore09 Technically you could use the DB API to Delete it in the test case. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

aslakknutsen commented:
*[workitemtype.go, line 69 \[r6\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmo-r6-69:-KQ0HXfkKI01qNE_N3xj:b-vq68x6) ([raw file](https://github.com/almighty/almighty-core/blob/6594b636cf10dce5a632756442d824a8350645dc/workitemtype.go#L69)):*
> Ok, one more related question. "API for Update WIT". What to do with existing WIs of that type. Do we have workflow for that ?

Not really... :)

Nor do we have Experience defined for the whole WIT Extension stuff...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

pranavgore09 commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[workitemtype.go, line 69 \[r6\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmo-r6-69:-KQ0M2YjsG_NHhJbmjCw:b4yxde) ([raw file](https://github.com/almighty/almighty-core/blob/6594b636cf10dce5a632756442d824a8350645dc/workitemtype.go#L69)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> > Ok, one more related question. "API for Update WIT". What to do with existing WIs of that type. Do we have workflow for that ?
> 
> Not really... :)
> 
> Nor do we have Experience defined for the whole WIT Extension stuff...
</details>
Yes, so as of now, action items for me boils down to -
1. remove DELETE API completely
2. add tests for create, fetch WIT only
3. Do not proceed with Update API

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

pranavgore09 commented:
Review status: 4 of 22 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[workitemtype.go, line 69 \[r6\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmo-r6-69:-KQLATebrW9sbhX6Gpez:b-ec9uhm) ([raw file](https://github.com/almighty/almighty-core/blob/6594b636cf10dce5a632756442d824a8350645dc/workitemtype.go#L69)):*
<details><summary><i>Previously, pranavgore09 (Pranav Gore) wrote…</i></summary>
> Yes, so as of now, action items for me boils down to -
> 1. remove DELETE API completely
> 2. add tests for create, fetch WIT only
> 3. Do not proceed with Update API
</details>
I have done with above three action items, I hope this branch will be merge-able. :smiley: considering create WIT stuff.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

pranavgore09 commented:
@aslakknutsen @tsmaeder : this merge request has Create and List API for WIT.
Changes made by @tsmaeder @kwk and me are merged in this branch.
Let us know if it is merge-able.

kwk commented:
Reviewed 25 of 25 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[workitemtype_test.go, line 27 \[r7\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KQP3rTIIisxR5la6whn:-KQP3rTIIisxR5la6who:be3d9k8) ([raw file](https://github.com/almighty/almighty-core/blob/942fe1fdea2a296e52cde4aeb9dd3ee11f0af20e/workitemtype_test.go#L27)):*
> ```Go
> //-----------------------------------------------------------------------------
> 
> // The WorkItemTypeTestSuite has state the is relevant to all tests.
> ```

My bad: "has state the is relevant" should be "has state that is relevant"

---

*[workitemtype_test.go, line 142 \[r7\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KQP4FmXaDMbLDMH5ifu:-KQP4FmXaDMbLDMH5ifv:b1hknd3) ([raw file](https://github.com/almighty/almighty-core/blob/942fe1fdea2a296e52cde4aeb9dd3ee11f0af20e/workitemtype_test.go#L142)):*
> ```Go
> func (s *WorkItemTypeSuite) createWorkItemTypePerson() (http.ResponseWriter, *app.WorkItemType) {
> 
> 	// Create the type for the "color" field
> ```

"color" should be "name". Sorry, my bad again.

---

*[models/workitemtype_repository.go, line 103 \[r7\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KQP5SAPrNnEWXMHRoiz:-KQP5SAPrNnEWXMHRoj-:bhvqtbf) ([raw file](https://github.com/almighty/almighty-core/blob/942fe1fdea2a296e52cde4aeb9dd3ee11f0af20e/models/workitemtype_repository.go#L103)):*
> ```go
> // List returns work item types selected by the given criteria.Expression, starting with start (zero-based) and returning at most "limit" item types
> ```

This comment needs to be changed to the following comment because the criteria part is not used.

```go
// List returns work item types starting with "start" (zero-based) and returns at most "limit" work item types.
```

---

*[alm, line 0 \[r7\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KQP3JbWzPPX3-5vdaBo:-KQP3JbWzPPX3-5vdaBp:b69r5y0) ([raw file](https://github.com/almighty/almighty-core/blob/942fe1fdea2a296e52cde4aeb9dd3ee11f0af20e/alm#L0)):*
This file must not be checked int.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

kwk commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149#-:-KQP6CXZ5hIbyPdcxb8Y:bnfp4nl)*
<!-- Sent from Reviewable.io -->

pranavgore09 commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[alm, line  \[r7\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KQP3JbWzPPX3-5vdaBo:-KQP7Y4bcg-V1atlngAE:brjqqie) ([raw file](https://github.com/almighty/almighty-core/blob/942fe1fdea2a296e52cde4aeb9dd3ee11f0af20e/alm#L)):*
<details><summary><i>Previously, kwk (Konrad Kleine) wrote…</i></summary>
> This file must not be checked int.
</details>
removed. thanks

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

tsmaeder commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[design/user_types.go, line 26 \[r5\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPrgzlzMkEEwldXnyVe:-KQPWPlhAfYJR0HvV_HN:b-n36cy5) ([raw file](https://github.com/almighty/almighty-core/blob/7f18c31ef9ff22d18b248de834db01f9ff354d04/design/user_types.go#L26)):*
<details><summary><i>Previously, pranavgore09 (Pranav Gore) wrote…</i></summary>
> yes, I will reuse fieldType.  But even after this I need to loop through the ctx.Payload.fields in order to make FieldDefinitions using required value. Because "required" attribute is in FieldDefinition and not in FieldType. And FieldType will not allow to have extra keys like "required". Little confused I am right now !!
</details>
As I understand it, you have to loop to dereference the pointer that goa generates for the "Type" field in the fieldDefinition struct. The attribute in the "fieldType" in media_types.go is marked as required, I don't understand why it generates a pointer field here. So tough luck, I guess.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149#-:-KQPXlhO6dTBtXHPcpZ_:bnfp4nl)*
<!-- Sent from Reviewable.io -->

pranavgore09 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149#-:-KQPd_spDW2_HSIfX5ND:bg7igkv)*
<!-- Sent from Reviewable.io -->

pranavgore09 commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[workitemtype_test.go, line 27 \[r7\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KQP3rTIIisxR5la6whn:-KQPeNOH1dkO4U-fKCm6:b4tdi24) ([raw file](https://github.com/almighty/almighty-core/blob/942fe1fdea2a296e52cde4aeb9dd3ee11f0af20e/workitemtype_test.go#L27)):*
<details><summary><i>Previously, kwk (Konrad Kleine) wrote…</i></summary>
> My bad: "has state the is relevant" should be "has state that is relevant"
</details>
For such comments also Reviewable wants me to take action ! Any help regarding this ? Or i should just ignore it.
It shows "To Reply" counter in red. :laughing: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

tsmaeder commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[workitemtype.go, line 69 \[r6\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmo-r6-69:-KQR-Z2UUKRfXl-tDHR4:bul6a25) ([raw file](https://github.com/almighty/almighty-core/blob/6594b636cf10dce5a632756442d824a8350645dc/workitemtype.go#L69)):*
<details><summary><i>Previously, pranavgore09 (Pranav Gore) wrote…</i></summary>
> I have done with above three action items, I hope this branch will be merge-able. :smiley: considering create WIT stuff.
</details>
My take would have been to remove delete from the JSON API, but have it in repository for testing purposes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

pranavgore09 commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[workitemtype.go, line 69 \[r6\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmo-r6-69:-KQRNlR7kki4ktQgCIUn:b-um0e00) ([raw file](https://github.com/almighty/almighty-core/blob/6594b636cf10dce5a632756442d824a8350645dc/workitemtype.go#L69)):*
<details><summary><i>Previously, tsmaeder (Thomas Mäder) wrote…</i></summary>
> My take would have been to remove delete from the JSON API, but have it in repository for testing purposes.
</details>
Yeah I also had similar thought. :smile:  But as of now I have deleted WIT using GORM raw API in test cases.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

kwk commented:
Reviewed 2 of 3 files at r9.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

kwk commented:
[test this]

aslakknutsen commented on design/user_types.go:
```diff
@@ -36,3 +36,10 @@ var UpdateWorkItemPayload = Type("UpdateWorkItemPayload", func() {
+var CreateWorkItemTypePayload = Type("CreateWorkItemTypePayload", func() {
+	Attribute("name", String, "Readable name of the type like Task, Issue, Bug, Epic etc.")
+	Attribute("fields", HashOf(String, fieldDefinition), "Type fields those must be followed by respective Work Items.")
+	Attribute("extendedTypeName", String, "If newly created type extends any existing type")
```
Don't expose this. It has no story atm.

tsmaeder commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[design/user_types.go, line 43 \[r9\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmn-r9-43:-KQVOGLtkxxL1kEBZI3D:b-tmiqdq) ([raw file](https://github.com/almighty/almighty-core/blob/e00128b9df6847f64c8d5d051da2f2a885039313/design/user_types.go#L43)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> Don't expose extendedTypeName. It has no story atm.
</details>
Not sure what you mean here...this is a necessary input to creating a type.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

aslakknutsen commented:
*[design/user_types.go, line 43 \[r9\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmn-r9-43:-KQVZZrXfuxf7ai8QT8Z:b-47spnu) ([raw file](https://github.com/almighty/almighty-core/blob/e00128b9df6847f64c8d5d051da2f2a885039313/design/user_types.go#L43)):*
<details><summary><i>Previously, tsmaeder (Thomas Mäder) wrote…</i></summary>
> Not sure what you mean here...this is a necessary input to creating a type.
</details>
Because it's a feature looking for an Experience at this time. WIT Extension/Mixin shouldn't have been implemented at all.  The Experience of inheritance etc could technically only be a UI feature. Core needs the complete WIT to operate anyway.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

tsmaeder commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[design/user_types.go, line 43 \[r9\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmn-r9-43:-KQVjhT_iAASdypixHEX:b-b7uv7k) ([raw file](https://github.com/almighty/almighty-core/blob/e00128b9df6847f64c8d5d051da2f2a885039313/design/user_types.go#L43)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> Because it's a feature looking for an Experience at this time. WIT Extension/Mixin shouldn't have been implemented at all.  The Experience of inheritance etc could technically only be a UI feature. Core needs the complete WIT to operate anyway.
</details>
Yes, that's why I don't think the "parent chain" should be returned when getting a work item type. However, inheritance is a absolutely basic concept for ensuring presence of basic fields and establishing base types.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

kwk commented:
[test this]

tsmaeder commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 25 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149#-:-KQVtrpe8HAA1OJv3nGN:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tsmaeder commented:
Reviewed 2 of 10 files at r6, 9 of 25 files at r7, 2 of 3 files at r9.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

aslakknutsen commented:
*[design/user_types.go, line 43 \[r9\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmn-r9-43:-KQW3ul-6Oqz6U1NDPXu:bz2mksp) ([raw file](https://github.com/almighty/almighty-core/blob/e00128b9df6847f64c8d5d051da2f2a885039313/design/user_types.go#L43)):*
<details><summary><i>Previously, tsmaeder (Thomas Mäder) wrote…</i></summary>
> Yes, that's why I don't think the "parent chain" should be returned when getting a work item type. However, inheritance is a absolutely basic concept for ensuring presence of basic fields and establishing base types.
</details>
If that's the case;

* Why is extendedTypeName optional in the REST API?
* Why is extendedTypeName optional in the DB API?
* How does having a non enforced extendedTypeName exposed in the User API ensure the presence of basic fields & base types?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

tsmaeder commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[design/user_types.go, line 43 \[r9\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmn-r9-43:-KQW5f2nBKSy5IgUxbQT:boorgd7) ([raw file](https://github.com/almighty/almighty-core/blob/e00128b9df6847f64c8d5d051da2f2a885039313/design/user_types.go#L43)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> If that's the case;
> 
> * Why is extendedTypeName optional in the REST API?
> * Why is extendedTypeName optional in the DB API?
> * How does having a non enforced extendedTypeName exposed in the User API ensure the presence of basic fields & base types?
</details>
It should be non-optional in the REST API, but optional internally for setting up the basic types

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

aslakknutsen commented:
*[design/user_types.go, line 43 \[r9\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmn-r9-43:-KQWA4Hwdq0UEkPRitWB:b-5rbued) ([raw file](https://github.com/almighty/almighty-core/blob/e00128b9df6847f64c8d5d051da2f2a885039313/design/user_types.go#L43)):*
<details><summary><i>Previously, tsmaeder (Thomas Mäder) wrote…</i></summary>
> It should be non-optional in the REST API, but optional internally for setting up the basic types
</details>
:+1: 

We're then saying a WorkitemType with ExtendedType is a 'full WorkItemType', and without ExtendedType it's a base type?

e.g. a Base Type could be 'Planner', for all objects that have the basic Fields required by the Planner to function 

I think that would imply that create WorkItem need to check that Type is not a Base Type, correct?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

pranavgore09 commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[design/user_types.go, line 43 \[r9\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmn-r9-43:-KQZ4C5nCR3khLhdKj4V:b-a4yzn6) ([raw file](https://github.com/almighty/almighty-core/blob/e00128b9df6847f64c8d5d051da2f2a885039313/design/user_types.go#L43)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> :+1: 
> 
> We're then saying a WorkitemType with ExtendedType is a 'full WorkItemType', and without ExtendedType it's a base type?
> 
> e.g. a Base Type could be 'Planner', for all objects that have the basic Fields required by the Planner to function 
> 
> I think that would imply that create WorkItem need to check that Type is not a Base Type, correct?
> 
> 
</details>
So I think extendedTypeName should be added to "Required" list and that's the only code change to be done.
wit_repository create method accepts nil extendedTypeName so that should be fine.

Also Can we define one term please, RootType==BaseType correct ? As we assign "/" as default value, I think RootType sounds better.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

pranavgore09 commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[design/user_types.go, line 43 \[r9\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmn-r9-43:-KQZ7nffKo4RQFqcnM4L:b7fwu64) ([raw file](https://github.com/almighty/almighty-core/blob/e00128b9df6847f64c8d5d051da2f2a885039313/design/user_types.go#L43)):*
<details><summary><i>Previously, pranavgore09 (Pranav Gore) wrote…</i></summary>
> So I think extendedTypeName should be added to "Required" list and that's the only code change to be done.
> wit_repository create method accepts nil extendedTypeName so that should be fine.
> 
> Also Can we define one term please, RootType==BaseType correct ? As we assign "/" as default value, I think RootType sounds better.
</details>
But if we do it non-optional in REST API what should be the value passed in by user if he/she wants to create RootType.
are we accepting nil or empty string ? in that case required flag does not fit well i think.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

pranavgore09 commented on main.go:
```diff
@@ -78,13 +78,20 @@ func main() {
+
 	c3 := NewWorkitemController(service, wiRepo, ts)
 	app.MountWorkitemController(service, c3)
 
```
@tsmaeder : Sorry to make a comment on Github and so late, but I want to understand that we already have mounted controllers for WI and WIT then do we need to do it again ? using c3 and c4 , Can I remove last two parts. I think it is redundancy. Let me know if i am missing anything.

tsmaeder commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[design/user_types.go, line 43 \[r9\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmn-r9-43:-KQZXdWG6wzfppzfAwXz:bon8jki) ([raw file](https://github.com/almighty/almighty-core/blob/e00128b9df6847f64c8d5d051da2f2a885039313/design/user_types.go#L43)):*
@pranav you can't create root types, that is a feature. 
@aslak 
> I think that would imply that create WorkItem need to check that Type is not a Base Type, correct?

I don't unerstand what you mean here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

tsmaeder commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[main.go, line 97 \[r9\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KQL9tuz5D0jJck4LNSZ-r9-97:-KQZYiBnT_jlY7dbqpj1:b-ugtlly) ([raw file](https://github.com/almighty/almighty-core/blob/e00128b9df6847f64c8d5d051da2f2a885039313/main.go#L97)):*
<details><summary><i>Previously, pranavgore09 (Pranav Gore) wrote…</i></summary>
> @tsmaeder : Sorry to make a comment on Github and so late, but I want to understand that we already have mounted controllers for WI and WIT then do we need to do it again ? using c3 and c4 , Can I remove last two parts. I think it is redundancy. Let me know if i am missing anything.
</details>
Yes, that looks like duplication.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

tsmaeder commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[main.go, line 97 \[r9\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KQL9tuz5D0jJck4LNSZ-r9-97:-KQZZ05ps6wi7RLnAAtE:b-h8mj5o) ([raw file](https://github.com/almighty/almighty-core/blob/e00128b9df6847f64c8d5d051da2f2a885039313/main.go#L97)):*
<details><summary><i>Previously, tsmaeder (Thomas Mäder) wrote…</i></summary>
> Yes, that looks like duplication.
</details>
remove c3 & c4, as the new names are better anyway.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

pranavgore09 commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[main.go, line 97 \[r9\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KQL9tuz5D0jJck4LNSZ-r9-97:-KQZhUQTOOR2B1X4z6GL:b-lhs69o) ([raw file](https://github.com/almighty/almighty-core/blob/e00128b9df6847f64c8d5d051da2f2a885039313/main.go#L97)):*
<details><summary><i>Previously, tsmaeder (Thomas Mäder) wrote…</i></summary>
> remove c3 & c4, as the new names are better anyway.
</details>
done, thanks

---

*[design/user_types.go, line 43 \[r9\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmn-r9-43:-KQZhb604lqhCv9umvBw:b-u7zr2d) ([raw file](https://github.com/almighty/almighty-core/blob/e00128b9df6847f64c8d5d051da2f2a885039313/design/user_types.go#L43)):*
<details><summary><i>Previously, tsmaeder (Thomas Mäder) wrote…</i></summary>
> @pranav you can't create root types, that is a feature. 
> @aslak 
> > I think that would imply that create WorkItem need to check that Type is not a Base Type, correct?
> 
> I don't unerstand what you mean here.
</details>
@thomas okay so, RootType != Basetype ? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

kwk commented:
Reviewed 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

aslakknutsen commented:
*[design/user_types.go, line 43 \[r9\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmn-r9-43:-KQ_ihZ3zUvhoGzGwdMy:b-6lu3dw) ([raw file](https://github.com/almighty/almighty-core/blob/e00128b9df6847f64c8d5d051da2f2a885039313/design/user_types.go#L43)):*
<details><summary><i>Previously, pranavgore09 (Pranav Gore) wrote…</i></summary>
> @thomas okay so, RootType != Basetype ? 
</details>
A BaseType is a way to enforce that the Item has certain predefined fields so Core can work. Still unsure why we expose this to the user, but ok..

@tsmaeder As far as I can tell from the code, I can create a WorkItem based on 'ANY' Type by simply providing the Type ID in the Create WorkItem request. 

Does it make sense to allow a User to create a WorkItem that is 'only' a BaseType?

Base WIT 1
WIT 2 extends 1

Can I Create a WI based on Type 1?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

pranavgore09 commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[design/user_types.go, line 43 \[r9\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmn-r9-43:-KQcgWn7Fy2W2JuOxPTg:byc7cl7) ([raw file](https://github.com/almighty/almighty-core/blob/e00128b9df6847f64c8d5d051da2f2a885039313/design/user_types.go#L43)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> A BaseType is a way to enforce that the Item has certain predefined fields so Core can work. Still unsure why we expose this to the user, but ok..
> 
> @tsmaeder As far as I can tell from the code, I can create a WorkItem based on 'ANY' Type by simply providing the Type ID in the Create WorkItem request. 
> 
> Does it make sense to allow a User to create a WorkItem that is 'only' a BaseType?
> 
> Base WIT 1
> WIT 2 extends 1
> 
> Can I Create a WI based on Type 1?
</details>
@aslak Technically yes user can create WI based on Type 1. But if we consider Project, Issue,Epic as a base type then it makes sense to make WI based on these.
What will be predefined fields for a base type ? any example so I can understand ?

@aslak, @thomas : Sorry to say this but can we please define scope for Tuesday(6th) release ? So I can make changes and push for review etc. If changes are major then related tests will also need touch up / fixes. Thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

aslakknutsen commented:
*[design/user_types.go, line 43 \[r9\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmn-r9-43:-KQeCPoXE93o5iTqkeWj:b-83xr70) ([raw file](https://github.com/almighty/almighty-core/blob/e00128b9df6847f64c8d5d051da2f2a885039313/design/user_types.go#L43)):*
<details><summary><i>Previously, pranavgore09 (Pranav Gore) wrote…</i></summary>
> @aslak Technically yes user can create WI based on Type 1. But if we consider Project, Issue,Epic as a base type then it makes sense to make WI based on these.
> What will be predefined fields for a base type ? any example so I can understand ?
> 
> @aslak, @thomas : Sorry to say this but can we please define scope for Tuesday(6th) release ? So I can make changes and push for review etc. If changes are major then related tests will also need touch up / fixes. Thanks.
</details>
@pranavgore09 The scope for Tuesday is defined here: https://github.com/almighty/almighty-core/issues/99 3 simple WITs with the Fields as mentioned. ID, Title, Description, State, Assignee. Whatever is the shortest sensible path to get there... 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

aslakknutsen commented:
*[design/user_types.go, line 43 \[r9\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmn-r9-43:-KQeEQB5QeQWAoqoSz1i:bu0fiv6) ([raw file](https://github.com/almighty/almighty-core/blob/e00128b9df6847f64c8d5d051da2f2a885039313/design/user_types.go#L43)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> @pranavgore09 The scope for Tuesday is defined here: https://github.com/almighty/almighty-core/issues/99 3 simple WITs with the Fields as mentioned. ID, Title, Description, State, Assignee. Whatever is the shortest sensible path to get there... 
</details>
@pranavgore09 I think a Base type is much simpler than an Epic or Issue. Those are 'final' things. At least the root "Planner" base would be the bare minimal we require to plan something; e.g. ID, Title, State, Assignee ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

pranavgore09 commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[design/user_types.go, line 43 \[r9\]](https://reviewable.io:443/reviews/almighty/almighty-core/149#-KPqfhlf6chozCtTKGmn-r9-43:-KQeIM-my1tEJ1po8mWI:b-8zus74) ([raw file](https://github.com/almighty/almighty-core/blob/e00128b9df6847f64c8d5d051da2f2a885039313/design/user_types.go#L43)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> @pranavgore09 I think a Base type is much simpler than an Epic or Issue. Those are 'final' things. At least the root "Planner" base would be the bare minimal we require to plan something; e.g. ID, Title, State, Assignee ?
</details>
thanks, 
added here https://github.com/almighty/almighty-core/issues/137#issuecomment-244332356

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149)*
<!-- Sent from Reviewable.io -->

kwk commented:
I don't want to be picky but this really needs to be merged into `121_create_wit` because there're many changes in master which will affect this PR as well (e.g. testing). And a lot of rebase work needs to be done once merged. @aslakknutsen @pranavgore09 @tsmaeder Please all say whether we can merge this to `121_create_wit` or not.

aslakknutsen commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149#-:-KQg2I_clEaLCtEUs4Ao:bg7igk9)*
<!-- Sent from Reviewable.io -->

pranavgore09 commented:
Thanks @aslak.
@konrad, yes for merge. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/149#-:-KQg2jRWP3JnKaJpmNET:b-52fz3t)*
<!-- Sent from Reviewable.io -->
